File: parcel-locker-cpp/library/src/managers/ClientManager.cpp
//
// Created by student on 10.06.2021.
//

#include <managers/ClientManager.h>
#include "model/Client.h"
#include "repositories/ClientRepository.h"
#include "exceptions/ClientManagerException.h"

ClientManager::ClientManager() {
    clientRepository = std::make_shared<ClientRepository>();
}

ClientPtr ClientManager::getClient(const std::string &telNumber) {
    try {
        if (telNumber.empty())
            throw ClientManagerException("telNumber variable is empty!");
    }
    catch (ClientManagerException &e) {
        std::cerr << e.what() << std::endl;
        throw;
    }
    return clientRepository->findByTelNumber(telNumber);
}

ClientPtr ClientManager::registerClient(const std::string &firstName, const std::string &lastName, const std::string &telNumber) {
    for(const auto& client : clientRepository->findAll()){
        if(client->getTelNumber() == telNumber)
        {
            return client;
        }
    }
    ClientPtr newClient = std::make_shared<Client>(firstName, lastName, telNumber);
    clientRepository->add(newClient);
    return newClient;
}

void ClientManager::unregisterClient(const ClientPtr &Client) {
    try {
        if(!Client)
            throw ClientManagerException("client is a nullptr!");
        if (this->getClient(Client->getTelNumber()) != nullptr)
            Client->setArchive(true);;
    }
    catch (ClientManagerException &e) {
        std::cerr << e.what() << std::endl;
        throw;
    }
}

File: parcel-locker/src/main/java/pl/pas/parcellocker/managers/ClientManager.java
//package pl.pas.parcellocker.managers;
//
//import pl.pas.parcellocker.exceptions.ClientManagerException;
//import pl.pas.parcellocker.model.Client;
//import pl.pas.parcellocker.repositories.ClientDao;
//
//import java.util.Arrays;
//
//public class ClientManager {
//
//    private final ClientDao clientRepository;
//
//    public ClientManager() {
//        clientRepository = new ClientDao();
//    }
//
//    public Client getClient(String telNumber) {
//        validateIfEmpty(telNumber);
//
//        return clientRepository.findByTelNumber(telNumber);
//    }
//
//    public Client registerClient(String firstName, String lastName, String telNumber) {
//        validateIfEmpty(firstName, lastName, telNumber);
//
//        for (Client client : clientRepository.findAll()) {
//            if (client.getTelNumber().equals(telNumber))
//                return client;
//        }
//
//        Client newClient = new Client(firstName, lastName, telNumber);
//        clientRepository.add(newClient);
//        return newClient;
//    }
//
//    public Client unregisterClient(Client client) {
//        if (client == null)
//            throw new ClientManagerException("client is a null!");
//
//        getClient(client.getTelNumber());
//
//        client.setActive(false);
//        clientRepository.update(client);
//        return client;
//    }
//
//    private void validateIfEmpty(String... values) {
//        if (Arrays.stream(values).anyMatch(String::isEmpty))
//            throw new ClientManagerException("Value is empty!");
//    }
//}
